From 5c1fc16e6b3eabdfacd54196b40b9063119855da Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:51:39 -0400
Subject: [tools] perf/ui/tui: Split help message for perf top and report

Message-id: <1408953210-26343-212-git-send-email-jolsa@redhat.com>
Patchwork-id: 88928
O-Subject: [PATCH RHEL7.1 BZ1133083 211/322] perf ui/tui: Split help message for perf top and report
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit e8e684a58b9bddde3fdb1a65cd26eb7a3e1e746e
Author: Namhyung Kim <namhyung.kim@lge.com>
Date: Thu Dec 26 14:37:58 2013 +0900

description
===========
Some hotkeys don't work for perf top so split help messages for them.

It'll be helpful to a future modification.  Also sort the message by
alphabetical order of the hotkey.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/ui/browsers/hists.c b/tools/perf/ui/browsers/hists.c
index a440e03..d43ec79 100644
--- a/tools/perf/ui/browsers/hists.c
+++ b/tools/perf/ui/browsers/hists.c
@@ -1400,6 +1400,35 @@ static int perf_evsel__hists_browse(struct perf_evsel *evsel, int nr_events,
  char script_opt[64];
  int delay_secs = hbt ? hbt->refresh : 0;
 
+#define HIST_BROWSER_HELP_COMMON     \
+ "h/?/F1        Show this window\n"    \
+ "UP/DOWN/PGUP\n"      \
+ "PGDN/SPACE    Navigate\n"     \
+ "q/ESC/CTRL+C  Exit browser\n\n"    \
+ "For multiple event sessions:\n\n"    \
+ "TAB/UNTAB     Switch events\n\n"    \
+ "For symbolic views (--sort has sym):\n\n"   \
+ "->            Zoom into DSO/Threads & Annotate current symbol\n" \
+ "<-            Zoom out\n"     \
+ "a             Annotate current symbol\n"   \
+ "C             Collapse all callchains\n"   \
+ "d             Zoom into current DSO\n"    \
+ "E             Expand all callchains\n"    \
+
+ /* help messages are sorted by lexical order of the hotkey */
+ const char report_help[] = HIST_BROWSER_HELP_COMMON
+ "P             Print histograms to perf.hist.N\n"
+ "r             Run available scripts\n"
+ "s             Switch to another data file in PWD\n"
+ "t             Zoom into current Thread\n"
+ "V             Verbose (DSO names in callchains, etc)\n"
+ "/             Filter symbol by name";
+ const char top_help[] = HIST_BROWSER_HELP_COMMON
+ "P             Print histograms to perf.hist.N\n"
+ "t             Zoom into current Thread\n"
+ "V             Verbose (DSO names in callchains, etc)\n"
+ "/             Filter symbol by name";
+
  if (browser == NULL)
   return -1;
 
@@ -1488,25 +1517,7 @@ static int perf_evsel__hists_browse(struct perf_evsel *evsel, int nr_events,
   case 'h':
   case '?':
    ui_browser__help_window(&browser->b,
-     "h/?/F1        Show this window\n"
-     "UP/DOWN/PGUP\n"
-     "PGDN/SPACE    Navigate\n"
-     "q/ESC/CTRL+C  Exit browser\n\n"
-     "For multiple event sessions:\n\n"
-     "TAB/UNTAB Switch events\n\n"
-     "For symbolic views (--sort has sym):\n\n"
-     "->            Zoom into DSO/Threads & Annotate current symbol\n"
-     "<-            Zoom out\n"
-     "a             Annotate current symbol\n"
-     "C             Collapse all callchains\n"
-     "E             Expand all callchains\n"
-     "d             Zoom into current DSO\n"
-     "t             Zoom into current Thread\n"
-     "r             Run available scripts('perf report' only)\n"
-     "s             Switch to another data file in PWD ('perf report' only)\n"
-     "P             Print histograms to perf.hist.N\n"
-     "V             Verbose (DSO names in callchains, etc)\n"
-     "/             Filter symbol by name");
+    is_report_browser(hbt) ? report_help : top_help);
    continue;
   case K_ENTER:
   case K_RIGHT:
-- 
1.7.1